## Exercise 4.2 Changing optimization parameters


In this exercise, you will use the same titanic dataset from the previous exercises. You can download the dataset from the following link:

[titanic_all_numeric.csv](https://drive.google.com/file/d/11nuYS-l3EXCsGJt81y4YTt3oTnFGaB68/view?usp=drive_link)

The data is pre-loaded into a pandas DataFrame called `df`. The `predictors` and `target` values are also pre-defined.

You'll want the optimization to start from scratch every time you change the learning rate, to give a fair comparison of how each learning rate did in your results. So we have created a function `get_new_model()` that creates an unoptimized model to optimize.

It's time to get your hands dirty with optimization. You'll now try optimizing a model at a very low learning rate, a very high learning rate, and a "just right" learning rate. You'll want to look at the results after running this exercise, remembering that a low value for the loss function is good.

## Instructions

* Import `SGD` from `tensorflow.keras.optimizers`.
* Create a list of learning rates to try optimizing with called `lr_to_test`. The learning rates in it should be `0.000001`, `0.01`, and `1.0`.
* Using a `for` loop to iterate over `lr_to_test`:
  * Use the `get_new_model()` function to build a new, unoptimized model.
  * Create an optimizer called `my_optimizer` using the `SGD()` constructor with keyword argument `learning_rate=lr`.
  * Compile your model. Set the optimizer parameter to be the SGD object you created above, and because this is a classification problem, use `'categorical_crossentropy'` for the loss parameter, , and `metrics=['accuracy']` to see the accuracy at the end of each epoch.
  * Fit the model using the `predictors` and the `target`.

## Code

Load data and convert the data to NumPy array:

In [ ]:
import numpy as np
import pandas as pd
from tensorflow.keras.utils import to_categorical

# Load csv file into the dataframe: df
df = pd.read_csv("titanic_all_numeric.csv")

# Convert the boolean values of the 'age_was_missing' column to integer
df.age_was_missing = df.age_was_missing.replace({True: 1, False: 0})

# Create predictors NumPy array: predictors
predictors = df.drop(['survived'], axis=1).values

# Save the number of columns in predictors: n_cols
n_cols = predictors.shape[1]

# Convert the target to categorical: target
target = to_categorical(df['survived'])


/tmp/ipython-input-1757448447.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.age_was_missing = df.age_was_missing.replace({True: 1, False: 0})


Create a neural network for a classification task

In [ ]:
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential

def get_new_model():
  # Set up the model
  model = Sequential()
  model.add(Dense(32, activation='relu', input_shape=(n_cols,)))
  model.add(Dense(2, activation='softmax'))

  return model

Optimize the model with different learning rate:

In [ ]:
import numpy as np
import pandas as pd
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import SGD

# Load the dataset
df = pd.read_csv("titanic_all_numeric.csv")

# Convert the boolean values of the 'age_was_missing' column to integer
df.age_was_missing = df.age_was_missing.replace({True: 1, False: 0})

# Create predictors NumPy array
predictors = df.drop(['survived'], axis=1).values

# Save the number of columns in predictors
n_cols = predictors.shape[1]

# Convert the target to categorical
target = to_categorical(df['survived'])

# Function to create a new model
def get_new_model():
    model = Sequential()
    model.add(Dense(32, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(2, activation='softmax'))
    return model

# List of learning rates to test
lr_to_test = [0.000001, 0.01, 1.0]

# Loop over each learning rate to test
for lr in lr_to_test:
    print('\n\nTesting model with learning rate: %f\n' % lr)

    # Build a new model for each test
    model = get_new_model()

    # Create SGD optimizer with the specified learning rate
    my_optimizer = SGD(learning_rate=lr)

    # Compile the model
    model.compile(optimizer=my_optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

    # Fit the model
    model.fit(predictors, target, epochs=10, verbose=2)




Testing model with learning rate: 0.000001

Epoch 1/10


/tmp/ipython-input-1241465416.py:12: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.age_was_missing = df.age_was_missing.replace({True: 1, False: 0})
/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


28/28 - 1s - 25ms/step - accuracy: 0.3490 - loss: 4.9521
Epoch 2/10
28/28 - 0s - 7ms/step - accuracy: 0.3490 - loss: 4.9305
Epoch 3/10
28/28 - 0s - 5ms/step - accuracy: 0.3490 - loss: 4.9090
Epoch 4/10
28/28 - 0s - 5ms/step - accuracy: 0.3490 - loss: 4.8875
Epoch 5/10
28/28 - 0s - 5ms/step - accuracy: 0.3502 - loss: 4.8661
Epoch 6/10
28/28 - 0s - 3ms/step - accuracy: 0.3502 - loss: 4.8448
Epoch 7/10
28/28 - 0s - 3ms/step - accuracy: 0.3502 - loss: 4.8235
Epoch 8/10
28/28 - 0s - 3ms/step - accuracy: 0.3502 - loss: 4.8022
Epoch 9/10
28/28 - 0s - 5ms/step - accuracy: 0.3502 - loss: 4.7810
Epoch 10/10
28/28 - 0s - 3ms/step - accuracy: 0.3513 - loss: 4.7598


Testing model with learning rate: 0.010000

Epoch 1/10
28/28 - 0s - 17ms/step - accuracy: 0.5859 - loss: 2.5080
Epoch 2/10
28/28 - 0s - 5ms/step - accuracy: 0.6375 - loss: 0.9391
Epoch 3/10
28/28 - 0s - 5ms/step - accuracy: 0.6611 - loss: 0.6748
Epoch 4/10
28/28 - 0s - 5ms/step - accuracy: 0.6599 - loss: 0.6535
Epoch 5/10
28/28 - 0s - 

The ouput should be:

Testing model with learning rate: 0.000001

28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4074 - loss: 5.5751   


Testing model with learning rate: 0.010000

28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5747 - loss: 3.3378   


Testing model with learning rate: 1.000000

28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5404 - loss: 28697.6289
